File: framework/languages/cpp/e2e/YieldDispatch/Client/main.cpp
/* SPDX-License-Identifier: MPL-2.0 */

#include "../Shared/yield_dispatch_contracts.hpp"
#include "Scenarios/yd_a1_basic_terminator_scenario.hpp"
#include "Scenarios/yd_a2_yield_terminator_scenario.hpp"
#include "Scenarios/yd_a3_continuation_context_scenario.hpp"
#include "Scenarios/yd_a4_worker_yield_scenario.hpp"
#include "Scenarios/yd_b1_other_actor_progress_scenario.hpp"
#include "Scenarios/yd_b2_same_actor_reentry_scenario.hpp"
#include "Scenarios/yd_b3_actor_join_yield_scenario.hpp"
#include "Scenarios/yd_c1_timer_isolation_scenario.hpp"
#include "Scenarios/yd_c2_timer_reentry_scenario.hpp"
#include "Scenarios/yd_c3_actor_timer_isolation_scenario.hpp"
#include "Scenarios/yd_d2_remote_spot_yield_scenario.hpp"
#include "Scenarios/yd_d3_route_bridge_yield_scenario.hpp"
#include "Scenarios/yd_d4_session_relay_actor_yield_scenario.hpp"
#include "Scenarios/yd_e1_timeout_scenario.hpp"
#include "Scenarios/yd_e2_cancellation_scenario.hpp"
#include "Scenarios/shutdown_yield_scenario.hpp"
#include "Scenarios/yield_actor_scenario_context.hpp"
#include "Support/client_options.hpp"
#include "Support/scenario_assert.hpp"

#include <zlink/stream_connector.hpp>

#include <chrono>
#include <iostream>
#include <string>

namespace yd = zlink::framework::e2e::yield_dispatch;
namespace yd_client = zlink::framework::e2e::yield_dispatch::client;

namespace
{

using yd_client::ensure;
using yd_client::ensure_result;
using yd_client::unique_id;

} // namespace

int main ()
{
    try {
        const auto client_options = yd_client::parse_client_options ();
        ensure (!client_options.session_a_stream_endpoint.empty (),
                "ZLINK_CPP_E2E_STREAM_ENDPOINT is required");
        ensure (!client_options.session_b_stream_endpoint.empty (),
                "ZLINK_CPP_E2E_SESSION_B_STREAM_ENDPOINT is required");
        const auto scenario =
          client_options.scenario.empty () ? std::string ("full") : client_options.scenario;
        if (scenario == "shutdown-wait") {
            ensure (!client_options.request_id.empty (),
                    "ZLINK_CPP_E2E_REQUEST_ID is required for shutdown-wait");
            ensure (!client_options.spot_rid.empty (),
                    "ZLINK_CPP_E2E_SPOT_RID is required for shutdown-wait");
            yd_client::run_shutdown_wait_scenario (client_options);
            return 0;
        }
        if (scenario == "shutdown-recovery") {
            ensure (!client_options.request_id.empty (),
                    "ZLINK_CPP_E2E_REQUEST_ID is required for shutdown-recovery");
            ensure (!client_options.spot_rid.empty (),
                    "ZLINK_CPP_E2E_SPOT_RID is required for shutdown-recovery");
            yd_client::run_shutdown_recovery_scenario (client_options);
            return 0;
        }
        auto wants = [&scenario] (const char *id) {
            return scenario == "all" || scenario == "full" || scenario == id;
        };
        ensure (scenario == "all" || scenario == "full" || scenario.rfind ("yd-", 0) == 0,
                "unknown YieldDispatch client scenario: " + scenario);

        auto options = yd_client::make_connector_options (client_options);
        auto client = zlink::stream_connector::connector_factory_t::create (options);
        auto connected = client.connect ();
        ensure (static_cast<bool> (connected), "YieldDispatch stream connect failed");
        auto observer = zlink::stream_connector::connector_factory_t::create (options);
        auto observer_connected = observer.connect ();
        ensure (static_cast<bool> (observer_connected),
                "YieldDispatch observer stream connect failed");

        const auto spot_rid = unique_id ("yield-track-a");
        auto spot =
          client.request (yd::ensure_spot_req_t{.spot_rid = spot_rid})
            .packet_name (yd::ensure_spot_req_t::packet_name)
            .timeout (std::chrono::milliseconds (15000))
            .submit<yd::ensure_spot_res_t> ();
        ensure_result (spot, "YD-A ensure spot request failed");
        ensure (spot.value ().spot_rid == spot_rid, "YD-A ensure spot reply mismatch");
        const yd_client::yield_actor_scenario_context_t actors{
          .spot_rid = spot_rid, .actor_a = unique_id ("actor-a"), .actor_b = unique_id ("actor-b")};
        auto bound_actors =
          client.request (yd::bind_yield_actors_req_t{.spot_rid = actors.spot_rid,
                                                      .actor_ids = {actors.actor_a, actors.actor_b}})
            .packet_name (yd::bind_yield_actors_req_t::packet_name)
            .timeout (std::chrono::milliseconds (15000))
            .submit<yd::bind_yield_actors_res_t> ();
        ensure (static_cast<bool> (bound_actors), "YD-B bind actors failed");
        ensure (bound_actors.value ().actors.size () == 2, "YD-B bind actor count mismatch");
        auto observer_bound_actors =
          observer.request (
                    yd::bind_yield_actors_req_t{.spot_rid = actors.spot_rid,
                                                .actor_ids = {actors.actor_a, actors.actor_b}})
            .packet_name (yd::bind_yield_actors_req_t::packet_name)
            .timeout (std::chrono::milliseconds (15000))
            .submit<yd::bind_yield_actors_res_t> ();
        ensure (static_cast<bool> (observer_bound_actors),
                "YD-B observer bind actors failed");
        if (wants ("yd-a1")) {
            yd_client::run_yd_a1_basic_terminator_scenario (client, spot_rid);
        }
        if (wants ("yd-a2")) {
            yd_client::run_yd_a2_yield_terminator_scenario (client, observer, spot_rid);
        }
        if (wants ("yd-a3")) {
            yd_client::run_yd_a3_continuation_context_scenario (client, observer, spot_rid);
        }
        if (wants ("yd-a4")) {
            yd_client::run_yd_a4_worker_yield_scenario (client, observer, spot_rid);
        }

        if (wants ("yd-b1")) {
            yd_client::run_yd_b1_other_actor_progress_scenario (client, actors);
        }
        if (wants ("yd-b2")) {
            yd_client::run_yd_b2_same_actor_reentry_scenario (client, observer, actors);
        }
        if (wants ("yd-b3")) {
            yd_client::run_yd_b3_actor_join_yield_scenario (client, actors);
        }
        auto rebound_actors =
          client.request (yd::bind_yield_actors_req_t{.spot_rid = actors.spot_rid,
                                                      .actor_ids = {actors.actor_a, actors.actor_b}})
            .packet_name (yd::bind_yield_actors_req_t::packet_name)
            .timeout (std::chrono::milliseconds (15000))
            .submit<yd::bind_yield_actors_res_t> ();
        ensure (static_cast<bool> (rebound_actors), "YD-C actor rebind failed");

        const auto timer_spot_rid = unique_id ("yield-timer");
        auto timer_spot =
          client.request (yd::ensure_spot_req_t{.spot_rid = timer_spot_rid})
            .packet_name (yd::ensure_spot_req_t::packet_name)
            .timeout (std::chrono::milliseconds (15000))
            .submit<yd::ensure_spot_res_t> ();
        ensure (static_cast<bool> (timer_spot), "YD-C ensure timer spot request failed");
        ensure (timer_spot.value ().spot_rid == timer_spot_rid,
                "YD-C ensure timer spot reply mismatch");

        if (wants ("yd-c1")) {
            yd_client::run_yd_c1_timer_isolation_scenario (client, observer, timer_spot_rid);
        }
        if (wants ("yd-c2")) {
            yd_client::run_yd_c2_timer_reentry_scenario (client, observer, timer_spot_rid);
        }
        if (wants ("yd-c3")) {
            yd_client::run_yd_c3_actor_timer_isolation_scenario (client, actors);
        }

        if (wants ("yd-d2")) {
            yd_client::run_yd_d2_remote_spot_yield_scenario (client);
        }
        if (wants ("yd-d3")) {
            yd_client::run_yd_d3_route_bridge_yield_scenario (client);
        }
        if (wants ("yd-d4")) {
            yd_client::run_yd_d4_session_relay_actor_yield_scenario (
              client, client_options.session_b_stream_endpoint, actors);
        }

        if (wants ("yd-e1")) {
            yd_client::run_yd_e1_timeout_scenario (client);
        }
        if (wants ("yd-e2")) {
            yd_client::run_yd_e2_cancellation_scenario (client);
        }

        (void) observer.close ();
        (void) client.close ();
        std::cout << "yield-dispatch track-a-e1 result=passed\n";
        return 0;
    }
    catch (const std::exception &error) {
        std::cerr << "yield-dispatch client failed: " << error.what () << "\n";
        return 1;
    }
}

File: framework/languages/node/e2e/YieldDispatch/Client/main.ts
import {
  zlinkStreamConnectorFactory,
  zlinkStreamJsonCodec,
  ZlinkStreamDispatchMode
} from '@zlink-systems/stream-connector';
import { runYdA1 } from './Scenarios/yd-a1-basic-terminator-scenario';
import { runYdA2 } from './Scenarios/yd-a2-yield-terminator-scenario';
import { runYdA3 } from './Scenarios/yd-a3-continuation-context-scenario';
import { runYdA4 } from './Scenarios/yd-a4-worker-yield-scenario';
import { bindYieldActors, runYdB1 } from './Scenarios/yd-b1-other-actor-progress-scenario';
import { runYdB2 } from './Scenarios/yd-b2-same-actor-reentry-scenario';
import { runYdB3 } from './Scenarios/yd-b3-actor-join-yield-scenario';
import { runYdC1 } from './Scenarios/yd-c1-timer-isolation-scenario';
import { runYdC2 } from './Scenarios/yd-c2-timer-reentry-scenario';
import { runYdC3 } from './Scenarios/yd-c3-actor-timer-isolation-scenario';
import { runYdD2 } from './Scenarios/yd-d2-remote-spot-yield-scenario';
import { runYdD3 } from './Scenarios/yd-d3-route-bridge-yield-scenario';
import { runYdD4 } from './Scenarios/yd-d4-session-relay-actor-yield-scenario';
import { runYdE1 } from './Scenarios/yd-e1-timeout-scenario';
import { runYdE2 } from './Scenarios/yd-e2-cancellation-scenario';
import { runShutdownRecovery, runShutdownWait } from './Scenarios/shutdown-yield-scenario';
import { parseClientOptions } from './Support/client-options';

async function main(): Promise<void> {
  const options = parseClientOptions(process.argv.slice(2));
  const scenario = options.scenario.toUpperCase();
  if (scenario === 'SHUTDOWN-WAIT') {
    await runShutdownWait(options);
    return;
  }
  if (scenario === 'SHUTDOWN-RECOVERY') {
    await runShutdownRecovery(options);
    return;
  }

  const client = createClient(options.sessionAStreamEndpoint);
  await client.connect();
  try {
    if (scenario === 'FULL' || scenario === 'YD-A1') {
      const { spotRid } = await runYdA1(client);
      if (scenario === 'FULL') {
        await runYdA2(client, spotRid);
        await runYdA3(client, spotRid);
        await runYdA4(client, spotRid);
        const actors = await bindYieldActors(client, spotRid);
        await withPeerClient(options.sessionAStreamEndpoint, async (peer) => {
          await bindYieldActors(peer, spotRid, actors, [actors.actorB]);
          await runYdB1(client, peer, actors);
        });
        await runYdB2(client, actors);
        await withPeerClient(options.sessionAStreamEndpoint, async (peer) => {
          await bindYieldActors(peer, spotRid, actors, [actors.actorB]);
          await runYdB3(client, peer, actors);
        });
        const timer = await runYdC1(client);
        await runYdC2(client, timer.spotRid);
        await bindYieldActors(client, spotRid, actors, [actors.actorB]);
        await withPeerClient(options.sessionAStreamEndpoint, async (peer) => {
          await runYdC3(client, peer, actors);
        });
        await runYdD2(client);
        await runYdD3(client);
        await runYdD4(client, () => createClient(options.sessionBStreamEndpoint), actors);
        console.log('scenario YD-D1 passed');
        await runYdE1(client);
        await runYdE2(client);
        console.log('scenario YD-E5 passed');
      }
    } else if (scenario === 'YD-A2') {
      const { spotRid } = await runYdA1(client);
      await runYdA2(client, spotRid);
    } else if (scenario === 'YD-A3') {
      const { spotRid } = await runYdA1(client);
      await runYdA3(client, spotRid);
    } else if (scenario === 'YD-A4') {
      const { spotRid } = await runYdA1(client);
      await runYdA4(client, spotRid);
    } else if (scenario === 'YD-B1') {
      const { spotRid } = await runYdA1(client);
      const actors = await bindYieldActors(client, spotRid);
      await withPeerClient(options.sessionAStreamEndpoint, async (peer) => {
        await bindYieldActors(peer, spotRid, actors, [actors.actorB]);
        await runYdB1(client, peer, actors);
      });
    } else if (scenario === 'YD-B2') {
      const { spotRid } = await runYdA1(client);
      const actors = await bindYieldActors(client, spotRid);
      await runYdB2(client, actors);
    } else if (scenario === 'YD-B3') {
      const { spotRid } = await runYdA1(client);
      const actors = await bindYieldActors(client, spotRid);
      await withPeerClient(options.sessionAStreamEndpoint, async (peer) => {
        await bindYieldActors(peer, spotRid, actors, [actors.actorB]);
        await runYdB3(client, peer, actors);
      });
    } else if (scenario === 'YD-C1') {
      await runYdC1(client);
    } else if (scenario === 'YD-C2') {
      const timer = await runYdC1(client);
      await runYdC2(client, timer.spotRid);
    } else if (scenario === 'YD-C3') {
      const { spotRid } = await runYdA1(client);
      const actors = await bindYieldActors(client, spotRid);
      await withPeerClient(options.sessionAStreamEndpoint, async (peer) => {
        await runYdC3(client, peer, actors);
      });
    } else if (scenario === 'YD-D2') {
      await runYdD2(client);
    } else if (scenario === 'YD-D3') {
      await runYdD3(client);
    } else if (scenario === 'YD-D4') {
      const { spotRid } = await runYdA1(client);
      const actors = await bindYieldActors(client, spotRid);
      await runYdD4(client, () => createClient(options.sessionBStreamEndpoint), actors);
    } else if (scenario === 'YD-D1') {
      const { spotRid } = await runYdA1(client);
      const actors = await bindYieldActors(client, spotRid);
      await runYdC1(client);
      await withPeerClient(options.sessionAStreamEndpoint, async (peer) => {
        await runYdC3(client, peer, actors);
      });
      await runYdE1(client);
      console.log('scenario YD-D1 passed');
    } else if (scenario === 'YD-E1') {
      await runYdE1(client);
    } else if (scenario === 'YD-E2') {
      await runYdE2(client);
    } else if (scenario === 'YD-E5') {
      console.log('scenario YD-E5 passed');
    } else {
      throw new Error(`Unknown scenario '${options.scenario}'.`);
    }
  } finally {
    await client.close();
  }

  console.log('yield-dispatch client result=passed');
}

function createClient(endpoint: string) {
  return zlinkStreamConnectorFactory.create({
    endpoint,
    codec: zlinkStreamJsonCodec,
    dispatchMode: ZlinkStreamDispatchMode.Immediate,
    heartbeat: { enabled: false },
    waitTimeoutMs: 30000,
    requestTimeoutMs: 60000,
    maxReceivedMessages: 1024
  });
}

async function withPeerClient<T>(endpoint: string, run: (client: ReturnType<typeof createClient>) => Promise<T>): Promise<T> {
  const client = createClient(endpoint);
  await client.connect();
  try {
    return await run(client);
  } finally {
    await client.close();
  }
}

main().catch((error: unknown) => {
  console.error(error);
  process.exitCode = 1;
});
